magit-stgit.el

```diff
@@ -593,66 +593,70 @@ ask for confirmation before deleting."
 
 ;;;; magit-stgit-mail
 
-(magit-define-popup magit-stgit-mail-popup
-  "Popup console for StGit mail."
-  'magit-stgit-commands
+(transient-define-prefix magit-stgit-mail ()
+  "Send a set of patches by e-mail."
+  :value '("--git" "--auto-recipients")
   :man-page "stg-mail"
-  :switches '((?m "Generate an mbox file instead of sending" "--mbox")
-              (?g "Use git send-email" "--git" t)
-              (?e "Edit cover letter before send" "--edit-cover")
-              (?a "Auto-detect recipients for each patch" "--auto")
-              (?A "Auto-detect To, Cc and Bcc for all patches from cover"
-                  "--auto-recipients" t))
-  :options '((?o "Set file as cover message" "--cover="
-                 (lambda (prompt default) (read-file-name "Find file: " default)))
-             (?v "Add version to [PATCH ...]" "--version=")
-             (?p "Add prefix to [... PATCH ...]" "--prefix=")
-             (?t "Mail To" "--to=")
-             (?c "Mail Cc" "--cc=")
-             (?b "Mail Bcc:" "--bcc="))
-  :actions '((?m "Send" magit-stgit-mail)))
+  ["Arguments"
+   ("-m" "Generate an mbox file instead of sending" "--mbox")
+   ("-g" "Use git send-email" "--git")
+   ("-e" "Edit cover letter before sending" "--edit-cover")
+   ("-a" "Automatically Cc the patch signers" "--auto")
+   ("-A" "Auto-detect To, Cc and Bcc for all patches from cover"
+    "--auto-recipients")
+   ""
+   ("-o" "Set file as cover message" "--cover="
+    :reader (lambda (_prompt initial-input _history)
+              (read-file-name "Find file: " nil nil nil initial-input)))
+   ("-v" "Add version to [PATCH ...]" "--version=")
+   ("-p" "Add prefix to [... PATCH ...]" "--prefix=")
+   ("-t" "Mail To" "--to=")
+   ("-c" "Mail Cc" "--cc=")
+   ("-b" "Mail Bcc" "--bcc=")]
+  ["Actions"
+   ("m" "Send" magit-stgit--mail)])
+
+(defun magit-stgit--mail-recipients (&optional cover)
+  "Return a list of zero or one of each of `--to', `--cc' and
+`--bcc' arguments for `stg mail', with the values determined from
+the cover letter file COVER, or the current buffer if COVER is
+nil."
+  (let ((buffer (current-buffer))
+        (fields '()))
+    (with-temp-buffer
+      (if cover
+          (insert-file-contents cover)
+        (insert-buffer-substring buffer))
+      (goto-char (point-min))
+      (while (re-search-forward (rx (group (or "To" "Cc" "Bcc")) ":" (+ blank)
+                                    (group (* nonl)) (* blank) eol)
+                                nil t)
+        (let ((field (match-string 1))
+              (recipient (match-string 2)))
+          (when (string-match "<" recipient)
+            (setf (alist-get field fields nil nil #'equal) recipient)))))
```

Comment:
I see a lint that using `alist-get` demands a minimum emacs version of 25.1. whereas the current requirement is emacs 24.4.

Seems like upping the minimum emacs version to 25.1, which was released in 2016, should be reasonably uncontroversial.

That said, one thing I experimented with was whether removing the `dash` dependency would be feasible. If we were to set the minimum emacs to 27.1 (2020-08-10), then we could use the builtin `flatten-list` instead of `-flatten` and a `(flatten-list (mapcar ...))` instead of `-mapcat` to eliminate dash.

I'm comfortable with having emacs 27.1 as the minimum version if it enables removal of a dependency.

Reply:
Nice catch. Perhaps we should set up some Makefiles soon to make the compilation reproducible? I think at least a couple of times the byte-compiler didn't manage to catch some issues for me just because there's always a bunch of stuff loaded into my Emacs. I've noticed @tarsius already has a streamlined set of scripts that he uses in his packages, so maybe we should use something similar. :-)

I'm comfortable with bumping up the version as well, even as high as 27.1, as I use at least Emacs 28 on all of my machines. Given that we would like to target StGit 2.0 which is even more cutting edge, I don't think this would be too big of a deal, but perhaps some people might want to use an older version of Emacs while testing out StGit 2.0? @tarsius Given that you have a lot of experience as an Emacs package maintainer, do you have any thoughts on this (and also regarding the appropriate versions of Magit and Transient)?